sasl/sasl.go

```diff
@@ -42,3 +44,21 @@ type StateMachine interface {
 	// value will be true.
 	Next(ctx context.Context, challenge []byte) (done bool, response []byte, err error)
 }
+
+// Metadata contains additional data for performing SASL authentication.
+type Metadata struct {
+	Address string
+}
+
+// WithContext returns a copy of the context with associated Metadata.
+func (m *Metadata) WithContext(ctx context.Context) context.Context {
+	return context.WithValue(ctx, ctxKey{}, m)
+}
+
+// MetadataFromContext retrieves the Metadata from the context.
+func MetadataFromContext(ctx context.Context) *Metadata {
+	if m, ok := ctx.Value(ctxKey{}).(*Metadata); ok {
+		return m
+	}
+	return nil
```

Comment:
This can be simplified:
```suggestion
	m, _ := ctx.Value(ctxKey{}).(*Metadata)
	return m
```